vignettes/customizing_styler.Rmd

```diff
@@ -0,0 +1,268 @@
+---
+title: "Customizing styler"
+author: "Lorenz Walthert"
+date: "8/10/2017"
+output: rmarkdown::html_vignette
+vignette: >
+  %\VignetteIndexEntry{Customizing styler}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+This vignette gives a high-level overview about how styler works and how you
+can define your own style guide and format code according to it. 
+
+# How styler works
+
+There are three major steps that styler performs in order to style code:
+
+1. Create a abstract syntax tree (AST) from `utils::getParseData()` that
+   contains positional information of every token. We call
+   this a nested parse table. You can learn more about how that is done
+   exactly in the vignettes "Data Structures" and "Manipulating the nested 
+   parse table".
+2. Apply transformer functions at each level of the nested parse table. In 
+   particular, we create two columns `lag_newlines` and `spaces`, which contain
+   the number of line breaks before the token and the number of spaces 
+   after the token. These will be the columns that most of our transformer 
+   functions will modify. To go through the whole nested structure and apply
+   the transformer functions repeatedly on every level of nesting, we use
+   a visitor approach, i.e. a function that takes functions as arguments and
+   applies them to every level of nesting. 
+3. Serialize the nested parse table, that is, extract the terminal tokens from 
+   the nested parse table and add spaces and line breaks between them as 
+   specified in the nested parse table. 
+
+The argument transformers is, apart from the code to style, the key argument 
+of functions such as `style_text()` and friends. By default, it is created 
+via the argument `style`. The transformers are a named
+list of transformer functions and other arguments passed to styler. To use the
+default style guide of styler (the tidyverse style guide), call
+`tidyverse_style()` to get the list of the transformer functions. Let's quickly
+look at what those are.
+```{r, message = FALSE}
+library("styler")
+library("dplyr")
+names(tidyverse_style())
+str(tidyverse_style(), give.attr = FALSE, list.len = 3)
+```
+
+We note that there are different types of transformer functions. `filler` is
+initializing some variables in the nested parse table (so it is not actually a
+transformer), the other elements modify either spacing, line break or tokens.
+`use_raw_indention` is not a function, it is just an option. All transformer
+functions have a similar structure. Let's pick one and look at it:
+```{r}
+tidyverse_style()$space$remove_space_after_opening_paren
+```
+
+As the name says, this function removes spaces after the opening parenthesis. But
+how? Its input is a flat parse table. Since the visitor functions `pre_visit()`
+and `post_visit()` will go through all levels of nesting, we don't need to care
+about that - we just need a function that can be applied to a flat parse table,
```

Comment:
Well, technically it is still a nested parse table (potentially with no child). So it would not be wrong to just say "[...] Its input is a nested parse table and we want to apply transformer functions to the top level of that nested parse table [..]". Alternatively, we can call it PATOLON (Parse table AT One Level Of Nesting). :-)